File: v1/LLD/Design-Patterns/State/code/State.java
// State Pattern — a Document (Context) delegates publish() to whichever DocumentState is active.
// Compile: javac State.java
// Run:     java State

interface DocumentState {
    void publish(Document document);
    String name();
}

class DraftState implements DocumentState {
    public void publish(Document document) {
        System.out.println("Draft submitted for moderation.");
        document.setState(new ModerationState());
    }

    public String name() {
        return "Draft";
    }
}

class ModerationState implements DocumentState {
    public void publish(Document document) {
        System.out.println("Moderator approved — publishing document.");
        document.setState(new PublishedState());
    }

    public String name() {
        return "Moderation";
    }
}

class PublishedState implements DocumentState {
    public void publish(Document document) {
        System.out.println("Already published — publish() has no further effect.");
    }

    public String name() {
        return "Published";
    }
}

class Document {
    private DocumentState state = new DraftState();

    void setState(DocumentState state) {
        this.state = state;
    }

    void publish() {
        state.publish(this);
    }

    String currentState() {
        return state.name();
    }
}

public class State {
    public static void main(String[] args) {
        Document doc = new Document();

        System.out.println("Current state: " + doc.currentState());
        doc.publish();

        System.out.println("Current state: " + doc.currentState());
        doc.publish();

        System.out.println("Current state: " + doc.currentState());
        doc.publish();

        System.out.println("Current state: " + doc.currentState());
    }
}

File: v1/LLD/Design-Patterns/State/code/state.py
"""
State Pattern — a Document (Context) delegates publish() to whichever DocumentState is active.
Run: python state.py
"""

from abc import ABC, abstractmethod


class DocumentState(ABC):
    @abstractmethod
    def publish(self, document):
        ...

    @abstractmethod
    def name(self):
        ...


class DraftState(DocumentState):
    def publish(self, document):
        print("Draft submitted for moderation.")
        document.set_state(ModerationState())

    def name(self):
        return "Draft"


class ModerationState(DocumentState):
    def publish(self, document):
        print("Moderator approved — publishing document.")
        document.set_state(PublishedState())

    def name(self):
        return "Moderation"


class PublishedState(DocumentState):
    def publish(self, document):
        print("Already published — publish() has no further effect.")

    def name(self):
        return "Published"


class Document:
    def __init__(self):
        self.state = DraftState()

    def set_state(self, state: DocumentState):
        self.state = state

    def publish(self):
        self.state.publish(self)

    def current_state(self):
        return self.state.name()


if __name__ == "__main__":
    doc = Document()

    print(f"Current state: {doc.current_state()}")
    doc.publish()

    print(f"Current state: {doc.current_state()}")
    doc.publish()

    print(f"Current state: {doc.current_state()}")
    doc.publish()

    print(f"Current state: {doc.current_state()}")
